### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): darts and edges are injective for walks, length of paths and trails are bounded

### PR DESCRIPTION
**Pull request by henkkuli**
Provide theorems stating that `SimpleGraph.Walk.darts` and `SimpleGraph.Walk.edges` are injective functions, and that the length of `SimpleGraph.Walk.IsTrail` and `SimpleGraph.Walk.IsPath` is bounded for finite graphs.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 1]
```diff
@@ -178,6 +178,19 @@ theorem IsTrail.count_edges_eq_one [DecidableEq V] {u v : V} {p : G.Walk u v} (h
...
+theorem IsTrail.length_le_card_edgeFinset [Fintype G.edgeSet] {u v : V}
+    (w : G.Walk u v) {h : w.IsTrail} : w.length ≤ G.edgeFinset.card := by
+  classical
+  let edges := w.edges.toFinset
+  have : edges.card = w.length := length_edges _ ▸ List.toFinset_card_of_nodup h.edges_nodup
+  rw [← this]
+  have : edges ⊆ G.edgeFinset := by
+    intro e h
+    refine mem_edgeFinset.mpr ?_
+    apply SimpleGraph.Walk.edges_subset_edgeSet w
```
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:190]:
```suggestion
    apply w.edges_subset_edgeSet
```

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 2]
```diff
@@ -214,6 +227,16 @@ theorem IsPath.of_append_right {u v w : V} {p : G.Walk u v} {q : G.Walk v w}
...
 
+theorem IsPath.length_lt_card_V [Fintype V] {u v : V} (w : G.Walk u v) {h : w.IsPath} :
+    w.length < Fintype.card V := by
+  classical
+  apply Nat.lt_of_add_one_le
+  let support := w.support.toFinset
+  have : support.card = w.length + 1 :=
+    length_support _ ▸ List.toFinset_card_of_nodup h.support_nodup
+  rw [← this]
+  exact Finset.card_le_card (Finset.subset_univ support)
```
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:238]:
```suggestion
  exact Finset.card_le_card support.subset_univ
```

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 3]
```diff
@@ -713,6 +713,19 @@ theorem edges_nodup_of_support_nodup {u v : V} {p : G.Walk u v} (h : p.support.N
     simp only [edges_cons, support_cons, List.nodup_cons] at h ⊢
     exact ⟨fun h' => h.1 (fst_mem_support_of_mem_edges p' h'), ih h.2⟩
 
+theorem edges_injective {u v : V} : Function.Injective (Walk.edges : G.Walk u v → List (Sym2 V))
+  | .nil, .nil, _ => by simp
```
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:717]:
```suggestion
  | .nil, .nil, _ => rfl
```

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 4]
```diff
@@ -178,6 +178,19 @@ theorem IsTrail.count_edges_eq_one [DecidableEq V] {u v : V} {p : G.Walk u v} (h
     {e : Sym2 V} (he : e ∈ p.edges) : p.edges.count e = 1 :=
   List.count_eq_one_of_mem h.edges_nodup he
 
+theorem IsTrail.length_le_card_edgeFinset [Fintype G.edgeSet] {u v : V}
+    (w : G.Walk u v) {h : w.IsTrail} : w.length ≤ G.edgeFinset.card := by
```
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:182]:
```suggestion
    {w : G.Walk u v} (h : w.IsTrail) : w.length ≤ G.edgeFinset.card := by
```

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 5]
```diff
@@ -214,6 +227,16 @@ theorem IsPath.of_append_right {u v w : V} {p : G.Walk u v} {q : G.Walk v w}
   rw [reverse_append] at h
   apply h.of_append_left
 
+theorem IsPath.length_lt_card_V [Fintype V] {u v : V} (w : G.Walk u v) {h : w.IsPath} :
```
[the remark this answers is not in the document]
Comment (Rida-Hamadani) [reply Mathlib/Combinatorics/SimpleGraph/Path.lean:230]: `IsPath.length_lt_card_verts`

**Comment by henkkuli** [comment]
Thanks for the feedback! I applied the suggested changes.

Turns out that `IsPath.length_lt_card_V` already exists as `IsPath.length_lt`, so I removed that theorem. It was a last-minute change for symmetry with `IsTrail.length_le_card_edgeFinset`, and apparently I didn't do my due diligence well enough.

**Comment by henkkuli** [comment]
Is there still something I should do to get this merged, or is this just waiting for a second round of reviews?

**Comment by IvanRenison** [comment]
I'm not a maintainer, so I can't marge it, a maintainer has to merge it, so it's waiting. Any way, you should mark as resolved the suggestions that made.

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Review by bryangingechen — APPROVED** [review]
Thanks!
bors r+

**Pull request closed without merging**